frontend/src/metabase/query_builder/components/notebook/NotebookSteps/NotebookSteps.tsx

```diff
@@ -41,52 +43,42 @@ function NotebookSteps({
   updateQuestion,
   readOnly = false,
 }: NotebookStepsProps) {
+  const metadata = useSelector(getMetadata);
   const [openSteps, setOpenSteps] = useState<OpenSteps>(
     getInitialOpenSteps(question, readOnly),
   );
-  const [lastOpenedStep, setLastOpenedStep] = useState<string | null>(null);
+  const [lastOpenedStep, setLastOpenedStep] = useState<
+    INotebookStep["id"] | null
+  >(null);
 
   const steps = useMemo(() => {
     if (!question) {
       return [];
     }
-    return getQuestionSteps(question, openSteps);
-  }, [question, openSteps]);
+    return getQuestionSteps(question, metadata, openSteps);
+  }, [metadata, question, openSteps]);
 
-  const handleStepOpen = useCallback((id: string) => {
+  const handleStepOpen = useCallback((id: INotebookStep["id"]) => {
     setOpenSteps(openSteps => ({ ...openSteps, [id]: true }));
     setLastOpenedStep(id);
   }, []);
 
-  const handleStepClose = useCallback((id: string) => {
+  const handleStepClose = useCallback((id: INotebookStep["id"]) => {
     setOpenSteps(openSteps => ({ ...openSteps, [id]: false }));
     setLastOpenedStep(lastOpenedStep =>
       lastOpenedStep === id ? null : lastOpenedStep,
     );
   }, []);
 
   const handleQueryChange = useCallback(
-    async (step: INotebookStep, query: StructuredQuery | Query) => {
-      // Performs a query update with either metabase-lib v1 or v2
-      // The StructuredQuery block is temporary and will be removed
-      // once all the notebook steps are using metabase-lib v2
-      if (query instanceof StructuredQuery) {
-        const datasetQuery = query.datasetQuery();
-        const updatedQuery = step.update(datasetQuery);
-        await updateQuestion(updatedQuery.question());
-      } else {
-        const updatedLegacyQuery = Lib.toLegacyQuery(query);
-        const updatedQuestion = question.setDatasetQuery(updatedLegacyQuery);
-        const updatedQuery = updatedQuestion.query() as StructuredQuery;
-        const cleanQuestion = updatedQuery.cleanNesting().question();
-        await updateQuestion(cleanQuestion);
-      }
-
-      // mark the step as "closed" since we can assume
-      // it's been added or removed by the updateQuery
-      handleStepClose(step.id);
+    async (query: Query) => {
+      const updatedLegacyQuery = Lib.toLegacyQuery(query);
+      const updatedQuestion = question.setDatasetQuery(updatedLegacyQuery);
+      const updatedQuery = updatedQuestion.query() as StructuredQuery;
+      const cleanQuestion = updatedQuery.cleanNesting().question();
```

Comment:
`dropStageIfEmpty` should do the same as `cleanNesting`

Reply:
Do you really need to drop more than 1 stage at once? We discussed it with @snoe and I think it is never the case

Reply:
I'm not sure. `cleanNesting` worked recursively, that's why I did it like that.

Updated to try to drop only the updated stage: 88bfe40fa55f1d2c3f4fd050511dd80951680c09